Convert Connect settings menu div to nav

jacobherrington opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/master/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/master/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This change converts a div to a more semantic element and takes care of a TODO in Connect.

## Related Tickets & Documents

## QA Instructions, Screenshots, Recordings

There are no visible UI changes.

![image](https://user-images.githubusercontent.com/11466782/106936199-e7595c00-66e1-11eb-82e5-572b5b0ac1ab.png)


### UI accessibility concerns?

This doesn't cause any complaints from axe, but if there is a missed opportunity to make this more usable please call it out :smile: 

## Added tests?

- [x] Yes (well... fixed the ones it broke)
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed

![giphy](https://user-images.githubusercontent.com/11466782/106936393-1ec80880-66e2-11eb-9b74-7b3cae17ae26.gif)

Ridhwana reviewed (approved):
This seems good to me. :)

aitchiss commented on app/javascript/chat/channels.jsx:
```diff
@@ -80,19 +80,23 @@ export const Channels = ({
-          </a>
-        </div>
-      </div>
+      <nav className="chatchannels__config" aria-label="configuration menu">
+        <img
+          alt="configuration"
+          src={ConfigImage}
+          style={{ height: '18px' }}
+          aria-hidden="true"
+        />
+        <ul className="chatchannels__configmenu">
+          <li>
+            <a href="/settings">Settings</a>
+          </li>
+          <li>
+            <a href="/report-abuse">Report Abuse</a>
+          </li>
+        </ul>
+      </nav>
```
Although this change correctly changes the group to a `nav`, it's still not accessible to assistive technology users. The nav items are rendered as hidden, and the menu can't be activated without using a mouse. The result is now instead of this whole block of content being hidden from screen reader users, now we tell them there is a `nav`, but it's empty 😞

See below: 

![VoiceOver rotor showing a list including empty nav](https://user-images.githubusercontent.com/20773163/107018210-0c100b00-6798-11eb-9e1b-c88f7b52030b.gif)

aitchiss commented on app/javascript/chat/channels.jsx:
```diff
@@ -80,19 +80,23 @@ export const Channels = ({
-        </div>
-      </div>
+      <nav className="chatchannels__config" aria-label="configuration menu">
+        <img
+          alt="configuration"
+          src={ConfigImage}
+          style={{ height: '18px' }}
+          aria-hidden="true"
+        />
```
I'm not sure if there is a plan to cover the broader issues in another PR (just let me know if so!), but to make this component accessible, this should be contained in a button, and activate on click rather than hover.

That button should live outside of the `nav` element so that it can toggle the visibility of the whole nav rather than just the list of links (the result of not doing that is that we surface an empty navigation to screen reader users as per my other comment).

When a user opens the nav by keyboard, we should also be sending focus to the first link in the list / returning focus to the button when the menu closes. Marcy introduced a pattern for this in the top navigation (activated by profile image click) - it's in `app/javascript/topNavigation/utilities.js`

Like I say, happy if this comes under another issue, but just want to make sure it's all captured!

Ridhwana replied:
Learning so much about accessibility by just reading this comment 😍

jacobherrington replied:
Connect needs a lot of love in this respect. I'll see if I can tackle this in one PR, but if not, I'll definitely open an issue with your comments. Thanks!

aitchiss replied:
No worries - yeah there's a lot in there 😅
I guess in this case I can't decide if surfacing an empty nav is better than no nav at all, if you see what I mean 🙃  Totally get there's a lot to wade into in Connect though and if it makes sense to chop it into smaller pieces that's grand - all steps forward for sure!

aitchiss reviewed (changes requested):
Left some comments re the current accessibility of this component in case it's not being covered elsewhere

jacobherrington commented:
> this should be contained in a button, and activate on click rather than hover :heavy_check_mark: 

> That button should live outside of the nav element so that it can toggle the visibility of the whole nav :heavy_check_mark: 

> When a user opens the nav by keyboard, we should also be sending focus to the first link in the list / returning focus to the button when the menu closes :heavy_check_mark: 

@aitchiss I *believe* I addressed your concerns on this PR. I'm certainly no Preact expert, so feel free to critique how I approached these bugs. :smile:

nickytonline commented:
Is this the same menu as the one in #12250. I know it's in a different area, but wonder if it's reused there.

aitchiss commented on app/javascript/chat/configMenu.jsx:
```diff
@@ -0,0 +1,52 @@
+
+    return (
+      <div className="chatchannels__config">
+        <button
+          onClick={this.handleClick}
+          aria-label="toggle configuration menu"
+          style={{ backgroundImage: `url(${ConfigImage})` }}
+          ref={this.configMenuButton}
```
(refactor/blocking) - This is working grand (and such a big improvement for accessibility!), but it'd be even better if we add `aria-expanded` so screen reader users know whether the menu is currently open/closed (and then we can remove the 'toggle' from the aria-label):

```suggestion
        <button
          onClick={this.handleClick}
          aria-expanded={visible}
          aria-label="configuration navigation menu"
          style={{ backgroundImage: `url(${ConfigImage})` }}
          ref={this.configMenuButton}
```

aitchiss commented on app/assets/stylesheets/chat.scss:
```diff
@@ -552,8 +549,16 @@
   cursor: pointer;
 
-  img {
+  button {
     opacity: 0.6;
+    border: none;
+    height: 18px;
+    width: 18px;
+    background-size: cover;
+  }
```
(refactor/blocking) It looks like the button background is a little funky in Safari:

![screenshot of button showing a gray background behind the three dots](https://user-images.githubusercontent.com/20773163/107245639-48a56600-6a27-11eb-949d-36754ef7df34.png)

I think it can be fixed by just adding an explicit background color:

```suggestion
  button {
    opacity: 0.6;
    border: none;
    height: 18px;
    width: 18px;
    background-size: cover;
    background-color: transparent;
  }
```

aitchiss reviewed (changes requested):
Looking great! Just two wee changes suggested 👍

jacobherrington commented:
@nickytonline It doesn't look like it shares any code (at a glance), but the work done here looks like it'd directly translate to cleaning that up.

jacobherrington commented:
@aitchiss Thanks! Suggestions committed :smile: Learned some new things from this PR, you were super helpful!! :hugs:

aitchiss reviewed (approved):
Looks great!

I think there's a future optimisation possible here to allow users to close the menu with `Escape` or a click outside, but it's not behaviour that we had before so doesn't necessarily fall under the scope of this refactor.

Delighted by this accessibility boost 🚀

ludwiczakpawel reviewed (approved):
Just ship this bad boy!
